### ISSUE TITLE
OS X emacs frozen --- bad version of emacs

### ISSUE THREAD

**Issue by fommil** [ensime/ensime-emacs#220]
_From @betehess on August 17, 2015 22:59_

It is "Waiting for connection-info"...

I was having this issue from time to time. I would normally force kill emacs, go back into the project, restart ensime. With the latest update for ensime (today), it systematically freezes right away. I don't think I have anything special in my configuration. I am not sure where to look to provide a better feedback.

**Comment by fommil** [comment]
_From @betehess on August 17, 2015 23:0_

And Activity Monitor shows that Emacs is stuck at 100% CPU usage.

**Comment by fommil** [comment]
_From @rorygraves on August 18, 2015 5:54_

Hmm sorry about that!,

Can you let me know:

Platform (Windows/Mac etc)
Emacs version.
What is displayed in the inferior-ensime-buffer (assuming you can get to it)

Thanks

**Comment by fommil** [comment]
Also, if you can profile where that CPU is would be great. He you have an example project to point us at?

Workaround is the git clone and checkout a version that was working for you, publishLocal and delete the ENSIME classpath cache file in .emacs.d/ensime

**Comment by fommil** [comment]
_From @betehess on August 18, 2015 21:18_

Here are a few more information.

> Platform (Windows/Mac etc)

OS X Yosemite 10.10.5. I had installed emacs through homebrew as explained in http://www.emacswiki.org/emacs/EmacsForMacOS.

> Emacs version.

```
➜  ~  emacs -v
emacsclient 24.5
➜  ~  cat ~/.emacs
(custom-set-variables
 '(inhibit-startup-screen t))
(custom-set-faces
 )

;; http://www.rensselaer.org/dept/acs/rpinfo/common/Computing/Consulting/Software/Emacs/Hints/backup.html
(setq make-backup-files nil)

;; http://www.emacswiki.org/emacs/EmacsForMacOS#toc26
(when (eq system-type 'darwin)
  (setq ns-command-modifier 'meta)
  )

(tool-bar-mode -1)

(require 'package)
(add-to-list 'package-archives
	     '("melpa" . "http://melpa.milkbox.net/packages/") t)
(package-initialize)

(when (not package-archive-contents)
  (package-refresh-contents))

(require 'ensime)
(add-hook 'scala-mode-hook 'ensime-scala-mode-hook)

(setq scala-indent:use-javadoc-style t
      popup-complete-enabled-modes '(scala-mode)
      ensime-default-buffer-prefix "ENSIME-"
      scala-outline-popup-select 'closest)

;; flx and idof
;; https://github.com/lewang/flx
(require 'flx-ido)
(ido-mode 1)
(ido-everywhere 1)
(flx-ido-mode 1)
;; disable ido faces to see flx highlights.
(setq ido-enable-flex-matching t)
(setq ido-use-faces nil)
(global-set-key "\M-x" 'smex)

(require 'windmove)
(windmove-default-keybindings 'meta)

;; http://web-mode.org/
(require 'web-mode)
(add-to-list 'auto-mode-alist '("\\.html?\\'" . web-mode))
(defun my-web-mode-hook ()
  "Hooks for Web mode."
  (setq web-mode-markup-indent-offset 2)
  (setq web-mode-css-indent-offset 2)
  (setq web-mode-code-indent-offset 2)
  (setq-default indent-tabs-mode nil)
  (setq-default tab-width 2)

)
(add-hook 'web-mode-hook 'my-web-mode-hook)

;; https://github.com/bbatsov/projectile
(projectile-global-mode)
(persp-mode)
(require 'persp-projectile)

;; https://raw.githubusercontent.com/magit/magit/next/Documentation/RelNotes/1.4.0.txt
;;(setq magit-last-seen-setup-instructions "1.4.0")
```
> What is displayed in the inferior-ensime-buffer (assuming you can get to it)

The last thing I can see is "Waiting for connection-info". I cannot do anything from there. Is there a place where I can find logs or anything?

> you have an example project to point us at?

Here is how I can systematically reproduce the problem on my machine:

```
➜  foobarbaz git:(master) ✗ tree
.
├── build.sbt
└── src
    └── main
        └── scala
            └── foobar
                └── Foo.scala

4 directories, 2 files
➜  foobarbaz git:(master) ✗ cat build.sbt
name in ThisBuild := "foobar"

organization in ThisBuild := "foooo"

version in ThisBuild := "0.1-SNAPSHOT"

scalaVersion in ThisBuild := "2.11.7"

scalacOptions in ThisBuild ++= Seq("-feature", "-deprecation", "-unchecked")

lazy val root = project.in(file("."))

➜  foobarbaz git:(master) ✗ cat src/main/scala/foobar/Foo.scala
package foobar

trait Foo
➜  foobarbaz git:(master) ✗ sbt gen-ensime
```

Then I open emacs on `src/main/scala/foobar/Foo.scala` (that's important), then `M-x ensime`, then it freezes. Apparently it works fine when I first launch ensime from `build.sbt` and only then open `Foo.scala`.

I hope this is not happening only on my machine :-/

**Comment by fommil** [comment]
which version of the ENSIME emacs client are you using? Look in the `.emacs.d/elpa` folder. Update to the latest and try again.

**Comment by fommil** [comment]
btw, server logs are in the buffer called `*ensime-inferior-server*`. You can also cancel any emacs hanging by typing `C-g` (this is a standard "quit" command in emacs). If you enable `debug-on-error` and `debug-on-quit` then you'll get some lisp stack traces.

It's definitely not happening for me on Linux.

**Comment by fommil** [comment]
actually, the buffer will be called `ENSIME-server-...` for you (because you've copied some config from me :wink:)

I don't advise using `popup-complete-enabled-modes` because we have `company-mode` nowadays. Although I don't think that would be causing a conflict.

**Comment by betehess** [comment]
> which version of the ENSIME emacs client are you using?

I update it pretty regularly:

```
➜  ~  ls ~/.emacs.d/elpa | grep ensime
ensime-20150818.528
```

> You can also cancel any emacs hanging by typing C-g

`C-g` just doesn't work when it's frozen.

> It's definitely not happening for me on Linux.

I wish that was (still) an option for me :-)

**Comment by fommil** [comment]
@betehess I've moved the ticket, let's continue the discussion on the `ensime-emacs` ticket

**Comment by fommil** [comment]
that's very weird. I've had reports from others that it is working on OS X and that is the latest version.

Are you using any weird encodings or is your Emacs itself out of date?

**Comment by betehess** [comment]
Well, I have started using OSX only 4 months ago, so it cannot be older than that :-) And as you can see above, I am already on emacs 24.5. As far as I know, I do not have anything custom related to encoding.

**Comment by betehess** [comment]
Also, it *is* working when I do not open a `.scala` file before starting `ensime`.

**Comment by betehess** [comment]
Alright, I was able to reproduce the problem on a colleague's laptop (OS X as well): it works well on 24.4, not on 24.5.

**Comment by fommil** [comment]
I've heard a few twitterings about emacs version problems on mac os x (unrelated to ENSIME).

@noelwelsh @channingwalton anything you could advise @betehess on here?

**Comment by noelwelsh** [comment]
I experienced random Emacs crashes and freezes once a day or more (not strictly related to Ensime) when I was using the EmacsForOSX build. I switched to `emacs-mac` and the crashes have reduced to a level in keeping with other applications (once a month or so). You can install this build of Emacs with

``` bash
brew tap railwaycat/emacsmacport
brew install emacs-mac
```

**Comment by channingwalton** [comment]
+1 @noelwelsh

**Comment by fommil** [comment]
@betehess can you confirm that this is the issue?

**Comment by betehess** [comment]
Not sure what this got closed before I even got a chance to test :-/

Note that I installed my emacs through brew, while my colleague used http://emacsformacosx.com/.

**Comment by noelwelsh** [comment]
The default Homebrew install of Emacs (`brew install emacs`) is the same as the EmacsForOsX build AFAIK. What I refer to above is **not** this build, but a slightly modified one known as `emacs-mac`.

**Comment by betehess** [comment]
@fommil I have the exact same issue with `emacs-mac`.

What I did:

```
➜ history | grep brew | tail -n 3
 4494  brew uninstall --force emacs
 4495  brew tap railwaycat/emacsmacport
 4496  brew install emacs-mac
```

@noelwelsh can you try to reproduce the scenario in https://github.com/ensime/ensime-emacs/issues/220#issuecomment-132363989? Note that it is important to start ensime while the scala file is opened.

**Comment by fommil** [comment]
@betehess can you try to reproduce with an open source project on github. It might be something to do with your project. Also, can you try with a completely minimal emacs config?

**Comment by fommil** [comment]
Btw, you should definitely be able to `C-g` the `waiting for connection-info` message.

Can you look in your ENSIME buffer? There might be logs there that are relevant.

Also, try deleting `.ensime_cache` and try again.

**Comment by fommil** [comment]
if `C-g` doesn't work try `ESC ESC ESC`

**Comment by noelwelsh** [comment]
@betehess: I use spacemacs, which handles running ensime for me. I don't actually know how to use it outside of spacemacs.

**Comment by fommil** [comment]
ok, well if you can reproduce on your friend's machine that's concerning. I really need to see the server logs to be able to tell you anything, are you sure that `C-g` doesn't even work to quit the ENSIME startup? If that doesn't work then it implies that emacs itself is broken, not ensime.

You can pass in a custom server logback file with the variable `ensime-server-logback` https://github.com/ensime/ensime-emacs/blob/master/ensime-vars.el#L90

Can you please try again after downloading https://github.com/ensime/ensime-emacs/blob/master/test/logback.xml (which will save the logs to disk as well as the buffer) with the setting

```elisp
(setq ensime-server-logback "/path/to/logback.xml")
```

That'll help a lot because I can't reproduce.

**Comment by peterbecich** [comment]
Emacs freezes when I try to start a second Ensime server.
I've tested this with two SBT projects, each with a freshly generated `.ensime` and `.ensime_cache`.  Starting the server in one of the projects goes fine, and Ensime mode connects to it successfully.  `M-x ensime` in the other project leads to `Waiting for connection-info`.
Emacs version 24.5.1 on OS X Yosemite
Using Ensime version 20150819.1214
Thanks!!  I get a lot of utility out of Ensime.

**Comment by fommil** [comment]
@peterbecich can you move your comment to a separate ticket please? I'm not sure its the same thing.

**Comment by fommil** [comment]
oh, you've solved it. Yeah, "waiting for connection-info" is a blocking call to wait for the server to come up. It can take a while first time. We'll make it async when we move to websockets. Or somebody could have a look and make it async now by putting in some "is initialised" predicate around all calls.

**Comment by fommil** [comment]
@peterbecich you shouldn't ever need to delete the .ensime files unless you've actually changed the project  definition (and even then you don't need to touch the cache)

**Comment by peterbecich** [comment]
@fommil , thanks.  I think I got in the habit of deleting `.ensime_cache` after reading your comment here (https://github.com/ensime/ensime-server/issues/670) to delete the Ivy cache.  Later I figured out that referred to `.ivy2` in the home directory.  Thanks!

**Comment by wilkenstein** [comment]
This issue is happening reproducibly every time for me with OS X 10.10.5, emacs 24.5 (9.0). It happens when I work in multiple projects -- a very common use-case for me. I'm using `sbt`.

The steps that reproduce it for me are
1. invoke `M-x ensime` in a `.scala` file in one project;
2. do work in that project;
3. invoke `M-x ensime` in a `.scala` file in a different project.

After Step 3, `Waiting for connection-info` displays in emacs, emacs goes on to use 100% CPU, and it is impossible to interact with the process. I have to force kill it.

The really odd thing is that, once this condition occurs for me, `M-x ensime` will no longer work in _any_ project. Invoking it from anywhere within a project hangs emacs with the `Waiting for connection-info`.

However, if I invoke `M-x ensime` outside of any project, giving it the project's `.ensime` file, ensime works great! I can load multiple `.ensime`s and everything works great.

I doubt this bug has anything to do with versions. I think there must be some subtle bug in ensime's operation when invoked within a project from one of the project's files....

**Comment by lunaryorn** [comment]
I can't give an exact recipe, but I also suffer from this frequently.  And likewise, things work fine when starting Ensime from the the SBT file.

**Comment by fommil** [comment]
Since it's only happening on OS XX I'm afraid you'll have to dig into this one yourselves. We can give guidance when you hit a brick wall but I'd suggest studying the codepath from starting ENSIME.

**Comment by channingwalton** [comment]
I've just this again and can't reproduce it now. I can switch between projects and no freezes.
I'll keep trying.

**Comment by xoltar** [comment]
This happens for me on linux (Ubuntu 14.04), with latest spacemacs.

**Comment by fommil** [comment]
@xoltar please add your comments to the similar Ubuntu thread and try to give us a step by step reproduction of the problem. I'm afraid I cannot reproduce.

**Comment by xoltar** [comment]
Ah, my apologies, it looks like I did (somehow?) have multiple copies of emacs running. I don't get this problem after killing and restarting.

I do, however, still get a complete freeze several seconds after the server starts.

Does that sound like an existing ticket, or should I open a new one?

**Comment by avernet** [comment]
+1

After upgrading my version of ENSIME (I was on a really ancient version), I'm also getting this on my project. Not much to add to what was said by others earlier: I'm on OS X, C-g or ESC ESC ESC have no effect, Emacs has to be killed; issue still happens after removing `.ensime_cache` right after starting Emacs, and I have no workaround (other than waiting for the next update to ENSIME which will fix this!).

**Comment by fommil** [comment]
We honestly have to idea what's causing this near can we reproduce. A hero from this thread will have to investigate the code. We're here to help, it's not too scary.

**Comment by channingwalton** [comment]
I am not sure where to begin TBH. Any pointers?

**Comment by fommil** [comment]
I think the PR from Aemon is getting close to something that we fundamentally broke and CI has been lying about. He may get somewhere so let's give him a few days.

**Comment by channingwalton** [comment]
Ok good.

**Comment by avernet** [comment]
Reiterating @betehess [comment](#issuecomment-132376989), which I missed on a first read of this thread, a **workaround** that seems to work in some, if not all, cases is to close all `.scala` files in Emacs before `M-x ensime`.

**Comment by fommil** [comment]
If that works for you, it is an extremely useful data point for us!

**Comment by aemoncannon** [comment]
Hey all. Just merged some fixes to the startup flow. All our tests are green now. Please try again and report back.

**Comment by avernet** [comment]
@aemoncannon This is most likely a *very* stupid question, but I am not seeing a new version on Melpa (I still see 20151102.207 as the latest version). Does it take a little while to get updated, or should we get it from somewhere else?

**Comment by fommil** [comment]
It'll appear eventually.

**Comment by channingwalton** [comment]
I'll try the update today

**Comment by wilkenstein** [comment]
I installed the latest ensime from MELPA today and tried the previous very reproducible for me way to hang emacs. It didn't hang!

I went into 3 separate scala projects, opened a file in each, and invoked M-x ensime with the file open. Did not hang and am not experience any issues so far.

Good job!

**Comment by fommil** [comment]
W00T!!!!!!!!!!

**Comment by avernet** [comment]
@aemoncannon Now working great. A-W-E-S-O-M-E!

**Comment by lunaryorn** [comment]
Seems to be gone for me, too.  @fommil @aemoncannon Thanks for your work!
